util/buildinfo/buildinfo.go

```diff
@@ -209,6 +209,36 @@ func decodeDeps(key string, attrs map[string]*string) (map[string]binfotypes.Bui
 	return res, nil
 }
 
+// dedupSources deduplicates main sources from dependencies ones.
+// FIXME: if a main source is also used in a dependency build, it will be removed. We should identify this case.
+func dedupSources(bi binfotypes.BuildInfo) binfotypes.BuildInfo {
```

Comment:
Dedup needs to be a recursive function. `dedupSources(bi *binfotypes.BuildInfo) ([]source)` from a dep and compare with current sources with a return value. If match then remove the source. If no match then add the current source to the return array. Might be easier if instead of `[]source` you have something like `map[source]struct{}` as long as source doesn't have slices and can be used as map key. If it can't then can use a hash function.